che-incubator/che-code
# Port @eclipse-che/theia-activity-tracker

This PR ports [eclipse-che-theia-activity-tracker](https://github.com/eclipse-che/che-theia/tree/main/extensions/eclipse-che-theia-activity-tracker) to che-code.

There were two main differences with https://github.com/eclipse-che/che-theia/tree/main/extensions/eclipse-che-theia-activity-tracker and the VS Code plugin introduced in this PR:

### Tracking
Instead of [front-end tracking like in che-theia](https://github.com/eclipse-che/che-theia/blob/60e8ad8784629294b0c088c039c4472567b11d87/extensions/eclipse-che-theia-activity-tracker/src/browser/che-theia-activity-tracker-contribution.ts#L36-L38), this PR tracks the following VS Code events using the extension API:
* `vscode.workspace.onDidChangeTextDocument`
* `vscode.window.onDidChangeActiveTextEditor`
* `vscode.window.onDidChangeTextEditorSelection`
* `vscode.window.onDidChangeTextEditorViewColumn`
* `vscode.window.onDidChangeWindowState`
* `vscode.window.onDidChangeTerminalState`
* `vscode.window.onDidChangeActiveTerminal`

### Determining the che-machine-exec url
Instead using an `EndpointService` to determine the machine-exec port like for [che-theia](https://github.com/eclipse-che/che-theia/blob/60e8ad8784629294b0c088c039c4472567b11d87/extensions/eclipse-che-theia-remote-impl-k8s/src/node/k8s-workspace-service-impl.ts#L79), this PR relies on an environment variable set by `entrypoint-volume.sh` which executes the machine-exec binary. The environment variable is read [here](https://github.com/che-incubator/che-code/blob/a954fdbd00f5a93e311be10d65ceae19b04cbf03/code/extensions/che-api/src/impl/k8s-workspace-service-impl.ts#L44).This was done because for Che-code, because for che-code, che-machine-exec is not running in its own container.

## To test this PR:
Since there is no idling set up for Che-code at the moment, to test this PR, create a workspace with this che-code image: `quay.io/dkwon17/che-code:idle`. Here is the editor devfile v2: https://gist.githubusercontent.com/dkwon17/1a96b69643260c9af23553f740132157/raw/72fb81012aa03618490b91cf887ff6a5cd12a889/gistfile1.txt.

For example, to start a golang che sample workspace, go to:
```
{CHE_HOST}/#https://github.com/che-samples/golang-example/tree/devfilev2?che-editor=https://gist.githubusercontent.com/dkwon17/1a96b69643260c9af23553f740132157/raw/72fb81012aa03618490b91cf887ff6a5cd12a889/gistfile1.txt
```

When the workspace is running, the `tools` container of the workspace pod should log `/activity/tick` `POST` messages when interacting with the che-code editor. Requests are sent once a minute at most to che-machine-exec:
<img width="1973" alt="image" src="https://user-images.githubusercontent.com/83611742/177221110-d51a8aed-581d-443e-ae63-0cab28112ede.png">

Signed-off-by: David Kwon <dakwon@redhat.com>
